src/stream/src/executor/managed_state/join/mod.rs

```diff
@@ -275,49 +233,52 @@ impl<K: HashKey, S: StateStore> JoinHashMap<K, S> {
     }
 
     /// Fetch cache from the state store. Should only be called if the key does not exist in memory.
-    async fn fetch_cached_state(&self, key: &K) -> RwResult<Option<JoinEntryState<S>>> {
-        let keyspace = self.get_state_keyspace(key)?;
-        JoinEntryState::with_cached_state(
-            keyspace,
-            self.data_types.clone(),
-            self.pk_data_types.clone(),
-            self.current_epoch,
-        )
-        .await
+    async fn fetch_cached_state(&self, key: &K) -> RwResult<Option<JoinEntryState>> {
+        let key = key.clone().deserialize(self.join_key_data_types.iter())?;
+        
+        let table_iter = self.state_table.iter_with_pk_prefix(key, self.current_epoch)?;
+        pin_mut!(table_iter);
+
+        let mut cached = BTreeMap::new();
+        #[for_await]
+        for row in table_iter {
+            let row = row?.into_owned();
+            let pk = row.by_indices(&self.pk_indices);
+
+            cached.insert(pk, JoinRow::from_row(row));
+        }
+        Ok(Some(JoinEntryState::with_cached(cached)))
     }
 
-    /// Create a [`JoinEntryState`] without cached state. Should only be called if the key
-    /// does not exist in memory or remote storage.
-    pub async fn init_without_cache(&mut self, key: &K) -> RwResult<()> {
-        let keyspace = self.get_state_keyspace(key)?;
-        let state = JoinEntryState::new(
-            keyspace,
-            self.data_types.clone(),
-            self.pk_data_types.clone(),
-        );
-        self.inner.put(key.clone(), state);
+    pub async fn flush(&mut self) -> RwResult<()> {
+        self.state_table.commit_with_value_meta(self.current_epoch).await.map_err(|e| RwError::from(e))
+    }
+
+    /// Insert a key
+    pub fn insert(&mut self, join_key: &K, pk: Row, value: JoinRow) -> RwResult<()>{
+        if let Some(entry) = self.inner.get_mut(join_key) {
+            entry.insert(pk, value.clone());
+        }
+        let key = join_key.clone().deserialize(self.join_key_data_types.iter())?;
+        // If no cache maintained, only update the flush buffer.
+        self.state_table.insert(&key, value.into_row())?;
```

Comment:
So the row count currently is not serialized, right?

Well the e2e tests shouldn't be able to pass without resolving this...? I guess we should decide on the way soon.

Reply:
The row count is currently added to the tail of a row as a datum as a workaround. :>

Reply:
What's your terminology here? row count eqauls to match degree in #2794 ?